### PR TITLE
[MOD-2328] iOS: Return upload responses to Ti.App.iOS.sessioncompleted event

### PR DESCRIPTION
**Pull request by rlustemberg**
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-24452

Version 2.1.0 of ti.urlsession has support for NSURLSessionUploadTask.
Once the NSURLRequest is completed, the sessioncompleted event is fired with a success status, because the request received it's response.
However, the response might have been a 401,still a success code would be set, because NSURLSessionUploadTask doesn't care  for HTTP error statuses.
We might also want to handle the response code after the upload task is completed.
To do this we only need to add a delegate method from NSURLSessionDataDelegate and modify the sessioncompleted delegate.

### CODE REVIEW

**Comment by hansemannn** [comment]
@rlustemberg Did some changes to clean-up the code and naming conventions. I think we should move the variable to the header file, easier to find.

**Comment by rlustemberg** [comment]
I'll move the variable to th header file and commit.

**Review by hansemannn — CHANGES_REQUESTED** [review]
One last thing: Please add the `responseText` property to the `Ti.App.iOS.sessioncompleted` event docs [here](https://github.com/appcelerator/titanium_mobile/blob/master/apidoc/Titanium/App/iOS/iOS.yml#L956). Like this:
```yml
      - name: responseText
        summary: The response text for upload tasks (since SDK 6.1.0).
        type: String
```

**Comment by rlustemberg** [comment]
Updating the docs right now

File: apidoc/Titanium/App/iOS/iOS.yml [hunk 1]
```diff
@@ -1,1256 +1,1260 @@
 ---
 name: Titanium.App.iOS
 summary: |
-    The top-level App iOS module, available only to iOS devices, that includes the facilities to
-    create and manage local notifications and background services.
+The top-level App iOS module, available only to iOS devices, that includes the facilities to
```
Comment (hansemannn) [inline apidoc/Titanium/App/iOS/iOS.yml]: now this is even worse :-) Let's do that after merging then, just restore the file.

**Comment by rlustemberg** [comment]
I'll fix this mess.

**Comment by rlustemberg** [comment]
Fixed now. I was copy and pasting while on Skype... bad

**Review by hansemannn — APPROVED** [review]

**Comment by hansemannn** [comment]
Approved!

**Comment by rlustemberg** [comment]
I thought this PR was already merged, so I opened a new ticket to return the response status code in the event object. I will add a new commit for this on the current PR

**Comment by hansemannn** [comment]
@rlustemberg The PR was waiting for QE-testing, so I'll remove the label and check the latest changes again.

File: iphone/Classes/TiApp.m [hunk 2]
```diff
@@ -799,17 +799,19 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didSend
 
 -(void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveData:(NSData *)data
 {
-    if (!uploadTaskResponses) {
-        uploadTaskResponses = [[[NSMutableDictionary alloc] init] retain];
+    if(!nsURLUploadTaskResponses){
+        nsURLUploadTaskResponses = [[NSMutableDictionary alloc] init];
```
Comment (hansemannn) [inline iphone/Classes/TiApp.m]: This changes overwrote the old changes we did (formatting, variable naming, etc.). Please restore that, I really want to include it soon..

File: iphone/Classes/TiApp.m [hunk 3]
```diff
@@ -799,17 +799,19 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didSend
...
     }
-	
-    NSMutableData *responseData =  [uploadTaskResponses objectForKey:@(dataTask.taskIdentifier)];
-    if (!responseData) {
-        responseData = [NSMutableData dataWithData:data];
-        [uploadTaskResponses setValue:responseData forKey:(NSString*)@(dataTask.taskIdentifier)];
+    NSMutableDictionary *responseObj =  [nsURLUploadTaskResponses objectForKey:@(dataTask.taskIdentifier)];
+    if (!responseObj) {
+
```
Comment (hansemannn) [inline iphone/Classes/TiApp.m]: Remove empty line

File: apidoc/Titanium/App/iOS/iOS.yml [hunk 4]
```diff
@@ -988,7 +988,10 @@ events:
 
       - name: responseText
         summary: The response text for upload tasks (since SDK 6.1.0).
-        type: String
+        type: Number
+
+      - name: statusCode
+        summary: The response status code for upload tasks (since SDK 6.1.0)
```
Comment (hansemannn) [inline apidoc/Titanium/App/iOS/iOS.yml]: Missing point. Also add `type: Number`

File: apidoc/Titanium/App/iOS/iOS.yml [hunk 5]
```diff
@@ -988,7 +988,10 @@ events:
 
       - name: responseText
         summary: The response text for upload tasks (since SDK 6.1.0).
-        type: String
+        type: Number
```
Comment (hansemannn) [inline apidoc/Titanium/App/iOS/iOS.yml]: The responseText should be a String, why Number?

File: iphone/Classes/TiApp.m [hunk 6]
```diff
@@ -797,33 +797,64 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didSend
     [[NSNotificationCenter defaultCenter] postNotificationName:kTiURLUploadProgress object:self userInfo:dict];
 }
 
+-(void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveData:(NSData *)data
+{
+    if(!nsURLUploadTaskResponses){
+        nsURLUploadTaskResponses = [[NSMutableDictionary alloc] init];
```
Comment (hansemannn) [inline iphone/Classes/TiApp.m]: How can this work? The property is named `uploadTaskResponses `. Please validate this again and rename + format properly.

File: iphone/Classes/TiApp.m [hunk 7]
```diff
@@ -797,33 +797,64 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didSend
...
+    }
+    NSMutableDictionary *responseObj =  [nsURLUploadTaskResponses objectForKey:@(dataTask.taskIdentifier)];
+    if (!responseObj) {
+        
+        NSMutableData * responseData = [NSMutableData dataWithData:data];
+        NSInteger statusCode = [(NSHTTPURLResponse *)[dataTask response] statusCode];
+        responseObj = [NSMutableDictionary dictionaryWithObjectsAndKeys: @(statusCode), @"statusCode",responseData, @"responseData", nil];
+        [nsURLUploadTaskResponses setValue:responseObj forKey:(NSString*)@(dataTask.taskIdentifier)];
+        
+    } else {
```
Comment (hansemannn) [inline iphone/Classes/TiApp.m]: Remove empty lines here.

File: iphone/Classes/TiApp.m [hunk 8]
```diff
@@ -797,33 +797,64 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didSend
...
     }
     
     if (error) {
         NSDictionary * errorinfo = [NSDictionary dictionaryWithObjectsAndKeys:NUMBOOL(NO), @"success",
-                                                NUMINTEGER([error code]), @"errorCode",
-                                                [error localizedDescription], @"message",
-                                                nil];
+                                    NUMINTEGER([error code]), @"errorCode",
+                                    [error localizedDescription], @"message",
+                                    nil];
```
Comment (hansemannn) [inline iphone/Classes/TiApp.m]: Restore indentation.

File: iphone/Classes/TiApp.m [hunk 9]
```diff
@@ -797,33 +797,64 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didSend
...
+            //we only send responseText as this is the responsesData dictionary only gets filled with data from uploads
+            responseText = [[NSString alloc] initWithData:[responseObj objectForKey:@"responseData"] encoding:NSUTF8StringEncoding];
+            statusCode = (NSInteger*)[responseObj objectForKey:@"statusCode"];
+            
+            [nsURLUploadTaskResponses removeObjectForKey:@(task.taskIdentifier)];
+        }
+        
+        NSDictionary * success = [NSMutableDictionary dictionaryWithObjectsAndKeys:NUMBOOL(YES), @"success",
+                                  NUMINT(0), @"errorCode",
+                                  @"", @"message",
```
Comment (hansemannn) [inline iphone/Classes/TiApp.m]: This is not good, an empty message is bad.

**Review by hansemannn — CHANGES_REQUESTED** [review]

File: iphone/Classes/TiApp.m (hunk 9 above)
Comment (rlustemberg) [inline iphone/Classes/TiApp.m]: I'll remove the key, it was like that already in the master branch

File: iphone/Classes/TiApp.m (hunk 6 above)
Comment (rlustemberg) [inline iphone/Classes/TiApp.m]: Sorry about that. I took working code from my own patch.

File: apidoc/Titanium/App/iOS/iOS.yml (hunk 4 above)
Comment (rlustemberg) [inline apidoc/Titanium/App/iOS/iOS.yml]: I thought this the type key was fixed on the last commit. I'm adding the point now

File: iphone/Classes/TiApp.m (hunk 2 above)
Comment (rlustemberg) [inline iphone/Classes/TiApp.m]: Just took care of it. Please let me know if you need me to fix something else. I'll take care of that asap.

File: iphone/Classes/TiApp.m [hunk 10]
```diff
@@ -780,30 +780,61 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didSend
...
 {
-  NSMutableDictionary *dict = [NSMutableDictionary dictionaryWithObjectsAndKeys:
-                                                       [NSNumber numberWithUnsignedInteger:task.taskIdentifier], @"taskIdentifier",
-                                                   nil];
+    if(!uploadTaskResponses){
+        uploadTaskResponses = [[NSMutableDictionary alloc] init];
+    }
+    NSMutableDictionary *responseObj =  [uploadTaskResponses objectForKey:@(dataTask.taskIdentifier)];
+    if (!responseObj) {
+        NSMutableData * responseData = [NSMutableData dataWithData:data];
```
Comment (hansemannn) [inline iphone/Classes/TiApp.m]: Why do you need a mutable copy here?

File: iphone/Classes/TiApp.m [hunk 11]
```diff
@@ -780,30 +780,61 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didSend
...
+        NSInteger statusCode = [(NSHTTPURLResponse *)[dataTask response] statusCode];
+        responseObj = [NSMutableDictionary dictionaryWithObjectsAndKeys: @(statusCode), @"statusCode",responseData, @"responseData", nil];
+        [uploadTaskResponses setValue:responseObj forKey:(NSString*)@(dataTask.taskIdentifier)];
+    } else {
+        [[responseObj objectForKey:@"responseData"] appendData:data ];
```
Comment (hansemannn) [inline iphone/Classes/TiApp.m]: Code-style: Remove space between `data ];`

File: iphone/Classes/TiApp.m [hunk 12]
```diff
@@ -780,30 +780,61 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didSend
...
 
-  if (session.configuration.identifier) {
-    [dict setObject:session.configuration.identifier forKey:@"sessionIdentifier"];
-  }
+-(void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error
+{
+    //FunctionName();
+    
+    NSMutableDictionary *dict = [NSMutableDictionary dictionaryWithObjectsAndKeys:
+                                 [NSNumber numberWithUnsignedInteger:task.taskIdentifier], @"taskIdentifier",
```
Comment (hansemannn) [inline iphone/Classes/TiApp.m]: Use `NUMUINTEGER()`

File: iphone/Classes/TiApp.m [hunk 13]
```diff
@@ -780,30 +780,61 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didSend
...
+    }
+    
+    if (error) {
+        NSDictionary * errorinfo = [NSDictionary dictionaryWithObjectsAndKeys:NUMBOOL(NO), @"success",
+                                    NUMINTEGER([error code]), @"errorCode",
+                                    [error localizedDescription], @"message",
+                                    nil];
+        [dict addEntriesFromDictionary:errorinfo];
+    } else {
+        NSMutableDictionary *responseObj = [uploadTaskResponses objectForKey:@(task.taskIdentifier)];
```
Comment (hansemannn) [inline iphone/Classes/TiApp.m]: Again, why mutable here? It does not mutate

**Review by hansemannn — CHANGES_REQUESTED** [review]
There are quite a few more changes required. Also the "since" references need to be update to 7.0.0. I'll try to take over the PR when our 6.3.0 / iPhone X release is out of the door and get it in for 7.0.0. Alrighty?

File: iphone/Classes/TiApp.m (hunk 13 above)
Comment (rlustemberg) [reply iphone/Classes/TiApp.m]: Actually there is another serious issue. I was reviewing this code today and found I was doing this
```
NSInteger *  statusCode = nil;
```
I've changed it to this
```
NSInteger  statusCode = 0;
```
For some reason this code was working allright , but today I tested for the first time with an iPhone 5C and it was causing a bad access exception when adding to the dictionary.

Comment (hansemannn) [reply iphone/Classes/TiApp.m]: well, you are basically making a pointer-object out of the underlaying primitive type. But assigning a primitive to the pointer then likely crashes it. Let me know when you fixed the other issues, then I do an additional commit to resolve other related issues.

File: iphone/Classes/TiApp.m (hunk 10 above)
Comment (rlustemberg) [reply iphone/Classes/TiApp.m]: I am not sure if I can modify the dictionary key itself. Which is necessary if the delegate method is called more than once and data needs to be appended.

**Comment by hansemannn** [comment]
@rlustemberg The PR needs to be rebased with master, you pushed from an outdated origin.

**Comment by rlustemberg** [comment]
Yes, I had merged before, deleted the commit, and I'm about to do the rebase.

**Comment by hansemannn** [comment]
Beautiful, thank you! We are using clang-format on master for a while, so the source stays properly formatted all the time. You can apply `clang-format -i -style=file Classes/TiApp.m` from the iphone/ directory to lint the source. You can install "clang-format" using brew as well! `brew install clang-format`

**Comment by rlustemberg** [comment]
Did you have a look at my reply regarding the NSMutableDictionary thing? Thanks a lot for the formatter tip! I'll start using it immediately.

**Review by hansemannn — APPROVED** [review]

**Comment by rlustemberg** [comment]
Done!

**Comment by hansemannn** [comment]
Looks good! Ready to merge once our CI completes all tests 🔥

**Pull request merged**